scipy/linalg/propack.py

```diff
@@ -0,0 +1,304 @@
+"""
+Python wrapper for PROPACK
+--------------------------
+
+PROPACK is a collection of Fortran routines for iterative computation
+of partial SVDs of large matrices or linear operators.
+
+Based on BSD licensed pypropack project:
+  http://github.com/jakevdp/pypropack
+  Author: Jake Vanderplas <vanderplas@astro.washington.edu>
+
+PROPACK source is BSD licensed, and available at
+  http://soi.stanford.edu/~rmunk/PROPACK/
+"""
+
+__all__ = ['svdp']
+
+import numpy as np
+from scipy.sparse.linalg import aslinearoperator
+
+from .misc import LinAlgError
+from ._propack import _spropack
+from ._propack import _dpropack
+from ._propack import _cpropack
+from ._propack import _zpropack
+
+
+_lansvd_dict = {
+    'f': _spropack.slansvd,
+    'd': _dpropack.dlansvd,
+    'F': _cpropack.clansvd,
+    'D': _zpropack.zlansvd,
+}
+
+
+_lansvd_irl_dict = {
+    'f': _spropack.slansvd_irl,
+    'd': _dpropack.dlansvd_irl,
+    'F': _cpropack.clansvd_irl,
+    'D': _zpropack.zlansvd_irl,
+}
+
+
+class _AProd:
+    """
+    Wrapper class for linear operator
+
+    The call signature of the __call__ method matches the callback of
+    the PROPACK routines.
+    """
+    def __init__(self, A):
+        try:
+            self.A = aslinearoperator(A)
+        except TypeError:
+            self.A = aslinearoperator(np.asarray(A))
+
+    def __call__(self, transa, m, n, x, y, sparm, iparm):
+        if transa == 'n':
+            y[:] = self.A.matvec(x)
+        else:
+            y[:] = self.A.rmatvec(x)
+
+    @property
+    def shape(self):
+        return self.A.shape
+
+    @property
+    def dtype(self):
+        try:
+            return self.A.dtype
+        except AttributeError:
+            return self.A.matvec(np.zeros(self.A.shape[1])).dtype
+
+
+def svdp(A, k, which='L', irl_mode=False, kmax=None,
```

Comment:
I'm guessing we'll want to put this is `scipy.sparse.linalg`. Please take a look `scipy.sparse.linalg.svds` and see whether it makes sense to include this as a `solver`.  You could add an `options` argument for additional options.

If it's a separate function, can you make sure it gets added to `__all__` and the docstring of `__init__.py` so we can see this documentation rendered?

It would be awesome if it can be hooked up to some of the the existing `scipy.sparse.linalg.svds` tests, in any case.

Reply:
Moved to `scipy.sparse.linalg` as a standalone solver as well as a solver for `scipy.sparse.linalg.svds`:
- standalone can do both `irl_mode` on/off
- svds solver can only do `irl_mode=True` to support `maxiter`

Also hooked up to the existing arpack svds test as a solver option.  It's failing on a few of them still and I'm not entirely sure why at this point.

I did import into `__init__.py`, there may be other things I need to do to get docs generated.